### DISCUSSION TITLE
Fbx exporter dependency to material

### DISCUSSION THREAD

**Discussion by etherealchain** [o3de/sig-graphics-audio#93] [Ideas]
In the current implementation, `ModelExporterComponent` export both model and material asset, which result in the following dependency:
```
FBX -> MaterialAsset -> MaterialType -> Shader
```
The result is whenever shader file is updated, fbx will be reprocessed by AssetProcessor, which wastes a lot of time and is not good for fast iteration. 

To address this issue, we need to find a way to break the dependency between fbx and material, a possible solution is to use the intermediate asset feature used by `MaterialTypeBuilder`.
To declare the product as intermediate, one has to set platform name to "common", and set the IntermediateAsset flag on the final product, to achieve that we may need to modify `SceneBuilderWorker` a bit.

After discussing with @santorac, the ideas are as follows:

1. When exporting material from a scene, export `MaterialSourceData` instead of material asset
2. Modify `SceneBuilderWorker` to output an extra "common" job, use this job to output `MaterialSourceData` and add the IntermediateAsset flag
3. The output job product `MaterialSourceData` will be put in AssetProcessorTemp folder
4. Because it's an intermediate product, it should be pickup by `MaterialBuilder` and finish the job

A side note is the model asset will still hold a material asset, as default material. So we have to load it when we are creating model asset, after the material had been built by `MaterialBuilder`.

**Comment by siliconvoodoo** [comment]
> A side note is the model asset will still hold a material asset, as default material.

Do you know why is that?
Is the proposal a way to change this:
![FBX-to-asset1](https://user-images.githubusercontent.com/25970839/201868859-b9a6cd11-784b-403e-98d6-3683148b4af5.svg)
to this:
![FBX-to-2-assets](https://user-images.githubusercontent.com/25970839/201868885-66a2adb1-a1d7-4b18-8fd7-5539ac25dfa1.svg)

**Reply by etherealchain** [reply]
Yeah, something like that.
Currently, both `MaterialAsset` and `ModelAsset` are final products of `ModelExporterComponent`.
Ideally, `MaterialSourceData` will be intermediate product, and `ModelAsset` will be the final product.
`MaterialSourceData` should be pickup by `MaterialBuilder` to build `MaterialAsset`.
Finally, the `MaterialAsset` should be used by the `ModelAsset` output from `ModelExporterComponent`, this way we can break the dependency.


> > A side note is the model asset will still hold a material asset, as default material.

> Do you know why is that?

I think model will hold material in material slot, it is by design right?

Let me update the graph, `ModelAsset` is not dependent on `MaterialAsset`, Fbx is.

**Reply by amzn-tommy** [reply]
The model asset has Asset<T> reference(s) to the default material asset(s) that are produced by the fbx. These are the materials that will be used to render the model if there is not material component or other system overriding the material for the model. But it's just a reference, so the data in the material asset itself should be able to change freely without needing the reference to be updated, thus the model itself shouldn't have to change or be re-processed even if the material asset needs to be re-processed.

**Reply by santorac** [reply]
The tricky thing is that the ModelAsset needs to have the AssetId of the final MaterialAsset, but the proposed design calls for the ModelAsset and MaterialSourceData to be produced at the same time. The MaterialAsset won't be produced until later, and that's when the AP will assign the material's AssetId. One theoretical approach would be for the exporter to "predict" the MaterialAsset's ID ahead of time, though I'm not sure this is possible. The other option would be to make ModelExporterComponet produce an _intermediate_ model file instead of the final ModelAsset. This intermediate file would be almost identical to the ModelAsset, except it would have a path to the intermediate MaterialSourceData, instead of the final AssetId. Then some new "ModelAssetFinalizerBuilder" would load the intermediate model file, look up the AssetId for each material, replace the intermediate paths with the AssetId, and then output the final ModelAsset.

**Reply by etherealchain** [reply]
The first option is impossible because the ID is generated randomly, unless we add a way to assign MaterialAsset ID when it is built by MaterialBuilder from MaterialSourceData.
The problem I can think of with the second option is, "ModelAssetFinalizerBuilder" will have to wait for MaterialBuilder to finish, which will still need dependency. But we have to avoid dependency here because we don't want Model to depend on Material.

**Reply by santorac** [reply]
That's a good observation. For the job dependency, we could use an "OrderOnce" dependency instead of "Order", which is intended for situations like this, where you need to discover the AssetId of something and after that you don't want the dependency to be applied. 

(I do wonder if there might be a better option though, I'll ask around.)

**Reply by etherealchain** [reply]
Ah yes, I see similar handling in `MaterialAssetDependenciesComponent::ReportJobDependencies`, I will give it a try.

**Reply by amzn-mike** [reply]
I want to point out that the UUID for the intermediate asset isn't random, it is based on the relative path of the file, so it is possible to get the UUID before AP has processed the file.  That said, I'm currently working on the metadata relocation system which will change how the UUID for intermediate assets is generated to be based on the hash of (SourceUUID:BuilderUuid:SubId).  All of this should be info available to the exporter.

**Reply by etherealchain** [reply]
Just want to be sure which part are you talking about.
Looks like the material asset id is generated randomly here
https://github.com/o3de/o3de/blob/40872d3dd7eeebfd8737099828a2a115a59f216f/Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp#L187
Am I missing something?

**Reply by santorac** [reply]
Ah I see, yes that is confusing. That random AssetId is not actually stored or used anywhere, other than a temporary ID in memory. The asset system requires every Asset object in memory to have an AssetId, so we just use a random one. The actual AssetId that the system assigns to the asset in the cache is different.

**Comment by santorac** [comment]
> FBX -> MaterialAsset -> MaterialType -> Shader

This dependency chain isn't quite correct. The FBX doesn't depend on a MaterialAsset, it depends on a MaterialTypeAsset. And it doesn't depend on any MaterialTypeAsset, it only depends on StandardPBR.materialtype. 

```
FBX -> MaterialType (StandardPBR.materialtype) -> Shader
```
